File: content/problems/1719-number-of-ways-to-reconstruct-a-tree/solutions/degree-parent-containment/Solution.kt
class Solution {
    fun checkWays(pairs:Array<IntArray>):Int{
        val graph=mutableMapOf<Int,MutableSet<Int>>();
        for(pair in pairs){
            graph.getOrPut(pair[0]){
                mutableSetOf()
            }.add(pair[1]);
            graph.getOrPut(pair[1]){
                mutableSetOf()
            }.add(pair[0])
        };
        val root=graph.keys.firstOrNull{
            graph[it]!!.size==graph.size-1
        }?:return 0;
        var answer=1;
        for(node in graph.keys){
            if(node==root)continue;
            val degree=graph[node]!!.size;
            var parent=-1;
            var parentDegree=Int.MAX_VALUE;
            for(neighbor in graph[node]!!){
                val candidateDegree=graph[neighbor]!!.size;
                if(candidateDegree>=degree&&candidateDegree<parentDegree){
                    parent=neighbor;
                    parentDegree=candidateDegree
                }
            };
            if(parent<0)return 0;
            for(neighbor in graph[node]!!)if(neighbor!=parent&&neighbor !in graph[parent]!!)return 0;
            if(parentDegree==degree)answer=2
        };
        return answer
    }
}

File: content/problems/1719-number-of-ways-to-reconstruct-a-tree/solutions/degree-parent-containment/solution.cpp
class Solution {
    public: int checkWays(vector<vector<int>>& pairs){
        unordered_map<int,unordered_set<int>> graph;
        for(auto& pair:pairs){
            graph[pair[0]].insert(pair[1]);
            graph[pair[1]].insert(pair[0]);
        }
        int root=-1;
        for(auto& [node,neighbors]:graph)if(neighbors.size()==graph.size()-1){
            root=node;
            break;
        }
        if(root<0)return 0;
        int answer=1;
        for(auto& [node,neighbors]:graph){
            if(node==root)continue;
            int degree=neighbors.size(),parent=-1,parentDegree=INT_MAX;
            for(int neighbor:neighbors){
                int candidateDegree=graph[neighbor].size();
                if(candidateDegree>=degree&&candidateDegree<parentDegree){
                    parent=neighbor;
                    parentDegree=candidateDegree;
                }
            }
            if(parent<0)return 0;
            for(int neighbor:neighbors)if(neighbor!=parent&&!graph[parent].count(neighbor))return 0;
            if(parentDegree==degree)answer=2;
        }
        return answer;
    }
};
